others/serialization.cpp

```diff
@@ -0,0 +1,163 @@
+ * binary file
+ */
+class Deserializer {
+ public:
+    /**
+     * Deserializes fundamental data types (like int, float, double, etc.) from
+     * a binary file.
+     * @tparam T The type of the data to be deserialized.
+     * @param in The input stream (std::ifstream).
+     * @param data The variable where the deserialized data will be stored.
+     *
+     * @note This function only works for fundamental types (primitives).
+     */
+    template <typename T>
+    static void deserialize(std::ifstream &in, T &data) {
+        static_assert(
+            std::is_fundamental<T>::value,
+            "Non-fundamental types are not allowed for this function!");
+        in.read(reinterpret_cast<char *>(&data), sizeof(T));
+    }
+
+    /**
+     * Deserializes a string from a binary file.
+     * @param in The input stream (std::ifstream).
+     * @param data The string where the deserialized content will be stored.
+     *
+     * The string is deserialized by reading its length, followed by the
+     * characters, and finally validating the delimiter throws error if the
+     * delimiter '|' is not found.
+     */
+    static void deserialize(std::ifstream &in, std::string &data) {
+        std::uint32_t length;
+        deserialize(in, length);  // Deserialize the length of the string.
+
+        if (length > 1024 * 1024)  // Sanity check to prevent huge strings.
+        {
+            throw std::runtime_error(
+                "Deserialized string length is too large.");
+        }
+
+        data.resize(length);  // Resize the string to fit the incoming data.
+        in.read(&data[0], length);  // Read the string characters.
+        char delimiter;
+        in.get(delimiter);  // Check the delimiter.
+        if (delimiter != '|') {
+            throw std::runtime_error("Delimiter '|' not found after string.");
+        }
+    }
+};
+/**
+ * @brief self test implementation
+ * @return void
+ */
+void tests() {
+    std::ofstream outFile("test_output.bin", std::ios::binary);
+    if (!outFile) {
+        std::cerr << "Error opening file for output.\n";
+        return;
+    }
+
+    int testInt = 12345;
+    double testDouble = 9876.54321;
+    char testChar = 'A';
+    std::string testString = "Testing String Serialization!";
+
+    // Serialize the data
+    Serializer::serialize(outFile, testInt);
+    Serializer::serialize(outFile, testDouble);
+    Serializer::serialize(outFile, testChar);
+    Serializer::serialize(outFile, testString);
+
+    outFile.close();
+
+    std::ifstream inFile("test_output.bin", std::ios::binary);
+    if (!inFile) {
+        std::cerr << "Error opening file for input.\n";
+        return;
+    }
+
+    int intResult;
+    double doubleResult;
+    char charResult;
+    std::string stringResult;
+
+    // Deserialize the data
+    Deserializer::deserialize(inFile, intResult);
+    Deserializer::deserialize(inFile, doubleResult);
+    Deserializer::deserialize(inFile, charResult);
```

Comment:
add test cases for the errors too via a try catch method and asserting that the error was indeed the same one

Reply:
I have added some try catch blocks in the tests function that will fail intentionally like:
- File not found
- Delimiter not found
- Writing strings larger than 1mb.
I guess there are okay?